File: src/statistics/drift.ts
/** Anytime-valid bounded-mean monitoring of randomly selected audit scenarios. */

import { ValidationError } from "../models.js";
import { mean } from "./dist.js";

export interface AuditRow {
  scenarioId: string; cluster: string; judgedSafe: boolean; auditSelected: boolean; auditProbability: number; humanReviewed: boolean;
  attempted?: Record<string, boolean>; occurred?: Record<string, boolean>; cost?: number | null;
}

export interface DriftResult {
  auditScenarios: number; auditClusters: number; familywiseAlpha: number; alarmThreshold: number; method: string; assumptions: string[];
  metrics: Record<string, { baseline: number; alarm: boolean; firstAlarmCluster: number | null; eValue: number; maximumEValue: number;
    trace: { cluster: string; scenarios: number; observedMean: number; logEValue: number }[] }>;
}

/**
 * A bounded Hoeffding e-process, applied once per completed top cluster.
 * Baseline keys: attempts:<harm>, occurrences:<harm>, or cost.
 */
export function detectDrift(auditRows: readonly AuditRow[], options: { baselines: Record<string, number>; alpha?: number; expectedShift?: number; costUpperBound?: number | null }): DriftResult {
  const { baselines, alpha = 0.05, expectedShift = 0.1, costUpperBound = null } = options;
  if (!(alpha > 0 && alpha < 1) || !(expectedShift > 0 && expectedShift <= 1) || !Object.keys(baselines).length) {
    throw new ValidationError("Specify baselines, alpha in (0,1), and normalized expectedShift in (0,1]");
  }
  if (auditRows.some((r) => !r.auditSelected || !r.judgedSafe || !r.humanReviewed)) {
    throw new ValidationError("Drift must use only randomly audited scenarios that operation judged safe");
  }
  const probabilities = new Set(auditRows.map((r) => Number(r.auditProbability ?? 0)));
  if (probabilities.size && (probabilities.size !== 1 || !((probabilities.values().next().value as number) > 0 && (probabilities.values().next().value as number) <= 1))) {
    throw new ValidationError("Audit rows require one common, positive random inclusion probability");
  }
  if (new Set(auditRows.map((r) => r.scenarioId)).size !== auditRows.length) throw new ValidationError("Operational audit scenarios may appear only once");
  const clustered = new Map<string, AuditRow[]>();
  const closed = new Set<string>();
  let previous: string | null = null;
  for (const row of auditRows) {
    if (previous !== null && row.cluster !== previous) closed.add(previous);
    if (closed.has(row.cluster)) throw new ValidationError("Completed audit clusters cannot be reopened");
    previous = row.cluster;
    const list = clustered.get(row.cluster) ?? [];
    list.push(row);
    clustered.set(row.cluster, list);
  }
  const threshold = Object.keys(baselines).length / alpha;
  const results: DriftResult["metrics"] = {};
  for (const [metric, baseline] of Object.entries(baselines)) {
    const upper = metric === "cost" ? Number(costUpperBound ?? 0) : 1;
    if (!(upper > 0) || !Number.isFinite(upper) || !Number.isFinite(baseline) || !(baseline >= 0 && baseline <= upper)) {
      throw new ValidationError("Baseline must be within finite bounds; cost requires a predeclared costUpperBound");
    }
    if (metric !== "cost" && !metric.startsWith("attempts:") && !metric.startsWith("occurrences:")) throw new ValidationError(`Unsupported audit drift metric: ${metric}`);
    const lambda = 4 * expectedShift;
    let logE = 0, maximum = 0;
    let alarm: number | null = null;
    const trace: DriftResult["metrics"][string]["trace"] = [];
    let index = 0;
    for (const [cluster, observations] of clustered) {
      index++;
      let values: number[];
      if (metric === "cost") {
        values = observations.map((r) => r.cost as number);
        if (values.some((v) => v === null || v === undefined || !Number.isFinite(v) || v < 0 || v > upper)) {
          throw new ValidationError("Every audited cost must be observed and within its predeclared bound");
        }
      } else {
        const [kind, harm] = [metric.slice(0, metric.indexOf(":")), metric.slice(metric.indexOf(":") + 1)];
        values = observations.map((r) => ((kind === "attempts" ? r.attempted : r.occurred)?.[harm] ? 1 : 0));
      }
      const m = mean(values) / upper;
      logE += lambda * (m - baseline / upper) - lambda ** 2 / 8;
      maximum = Math.max(maximum, logE);
      if (alarm === null && logE >= Math.log(threshold)) alarm = index;
      trace.push({ cluster, scenarios: observations.length, observedMean: m * upper, logEValue: logE });
    }
    results[metric] = { baseline, alarm: alarm !== null, firstAlarmCluster: alarm, eValue: Math.exp(Math.min(logE, 700)), maximumEValue: Math.exp(Math.min(maximum, 700)), trace };
  }
  return { auditScenarios: auditRows.length, auditClusters: clustered.size, metrics: results, familywiseAlpha: alpha, alarmThreshold: threshold,
    method: "one-sided bounded Hoeffding e-process with Bonferroni anytime threshold",
    assumptions: ["Audit inclusion is random and independent of outcomes among operationally safe scenarios.",
      "Each cluster is completed before it is added; future observations cannot be appended to an already monitored cluster.",
      "Under the null each next cluster's conditional mean is at most its predeclared baseline.",
      "The monitored population is the operationally safe book; escalated non-audit work is excluded.",
      "Clusters are equally weighted for drift; baselines must use the same cluster-level estimand."] };
}

File: src/signal_sdk/statistics/drift.py
"""Anytime-valid bounded-mean monitoring of randomly selected audit episodes."""

from __future__ import annotations

from collections import defaultdict
from collections.abc import Mapping, Sequence
from math import exp, log
from typing import Any

import numpy as np


def detect_drift(audit_rows: Sequence[Mapping[str, Any]], *, baselines: Mapping[str, float],
                 alpha: float = .05, expected_shift: float = .1,
                 cost_upper_bound: float | None = None) -> dict[str, Any]:
    """A bounded Hoeffding e-process, applied once per completed top cluster.

    baseline keys: attempts:<harm>, occurrences:<harm>, or cost. Every row
    must attest audit_selected=True and judged_safe=True. Selection must
    have a common nonzero inclusion probability independent of outcomes.
    """
    if not 0 < alpha < 1 or not 0 < expected_shift <= 1 or not baselines:
        raise ValueError("Specify baselines, alpha in (0,1), and normalized expected_shift in (0,1]")
    if any(not r.get("audit_selected") or not r.get("judged_safe") or not r.get("human_reviewed") for r in audit_rows):
        raise ValueError("Drift must use only randomly audited episodes that operation judged safe")
    probabilities = {float(r.get("audit_probability", 0)) for r in audit_rows}
    if probabilities and (len(probabilities) != 1 or not 0 < next(iter(probabilities)) <= 1):
        raise ValueError("Audit rows require one common, positive random inclusion probability")
    if len({r["episode_id"] for r in audit_rows}) != len(audit_rows):
        raise ValueError("Operational audit episodes may appear only once")
    clustered: dict[str, list[Mapping[str, Any]]] = defaultdict(list)
    closed = set()
    previous = None
    for row in audit_rows:
        if previous is not None and row["cluster"] != previous:
            closed.add(previous)
        if row["cluster"] in closed:
            raise ValueError("Completed audit clusters cannot be reopened")
        previous = row["cluster"]
        clustered[str(row["cluster"])].append(row)
    threshold = len(baselines) / alpha
    results = {}
    for metric, baseline in baselines.items():
        upper = float(cost_upper_bound or 0) if metric == "cost" else 1.
        if upper <= 0 or not np.isfinite(upper) or not np.isfinite(baseline) or not 0 <= baseline <= upper:
            raise ValueError("Baseline must be within finite bounds; cost requires a predeclared cost_upper_bound")
        if metric != "cost" and not metric.startswith(("attempts:", "occurrences:")):
            raise ValueError(f"Unsupported audit drift metric: {metric}")
        lam = 4 * expected_shift
        log_e = 0.
        maximum = 0.
        alarm = None
        trace = []
        for index, (cluster, observations) in enumerate(clustered.items(), 1):
            if metric == "cost":
                values = [r.get("cost") for r in observations]
                if any(v is None or not np.isfinite(v) or not 0 <= v <= upper for v in values):
                    raise ValueError("Every audited cost must be observed and within its predeclared bound")
            else:
                kind, harm = metric.split(":", 1)
                values = [float(bool(r.get("attempted" if kind == "attempts" else "occurred", {}).get(harm, False))) for r in observations]
            mean = float(np.mean(values)) / upper
            log_e += lam * (mean - baseline / upper) - lam ** 2 / 8
            maximum = max(maximum, log_e)
            if alarm is None and log_e >= log(threshold):
                alarm = index
            trace.append({"cluster": cluster, "episodes": len(observations), "observed_mean": mean * upper,
                          "log_e_value": log_e})
        results[metric] = {"baseline": baseline, "alarm": alarm is not None, "first_alarm_cluster": alarm,
                           "e_value": exp(min(log_e, 700)), "maximum_e_value": exp(min(maximum, 700)),
                           "trace": trace}
    return {"audit_episodes": len(audit_rows), "audit_clusters": len(clustered), "metrics": results,
            "familywise_alpha": alpha, "alarm_threshold": threshold,
            "method": "one-sided bounded Hoeffding e-process with Bonferroni anytime threshold",
            "assumptions": ["Audit inclusion is random and independent of outcomes among operationally safe episodes.",
                            "Each cluster is completed before it is added; future observations cannot be appended to an already monitored cluster.",
                            "Under the null each next cluster's conditional mean is at most its predeclared baseline.",
                            "The monitored population is the operationally safe book; escalated non-audit work is excluded.",
                            "Clusters are equally weighted for drift; baselines must use the same cluster-level estimand."]}
